internal/resourcepolicies/resource_policies_mgr.go

```diff
@@ -0,0 +1,139 @@
+	if err := mgr.updateResourcePolicies(refName, refType); err != nil {
+		return nil, errors.Wrap(err, "failed to update resource policies meta info")
+	}
+
+	resKey := resPoliciesKey{
+		refName: refName,
+		refType: refType,
+	}
+	if policies, exist := mgr.resPolicies[resKey]; exist {
+		return policies.policies, nil
+	}
+	return nil, fmt.Errorf("failed to get resource policies")
+}
+
+func (mgr *ResPoliciesManager) updateResourcePolicies(refName, refType string) error {
+	// TODO currently we only want to support type of configmap
+	if refType != "configmap" {
+		return fmt.Errorf("unsupported resource policies reference type %s", refType)
+	}
+
+	resKey := resPoliciesKey{
+		refName: refName,
+		refType: refType,
+	}
+	policiesConfigmap := &v1.ConfigMap{}
+	err := mgr.kbClient.Get(context.Background(), kbClient.ObjectKey{Namespace: mgr.namespace, Name: refName}, policiesConfigmap)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get resource policies %s/%s configmap", refName, mgr.namespace)
+	}
+	if policiesConfigmap == nil {
+		return errors.Wrapf(err, "get empty resource policies %s/%s configmap", refName, mgr.namespace)
+	}
+
+	updateResPolicies := func() error {
+		res, err := GetResourcePoliciesFromConfig(policiesConfigmap)
+		if err != nil {
+			return errors.Wrap(err, "failed to get the user resource policies config")
+		}
+		if res == nil {
+			return fmt.Errorf("%v resource policy not found", resKey)
+		}
+		mgr.resPolicies[resKey] = &resPoliciesMeta{
+			policies:                res,
+			configCreationTimestamp: policiesConfigmap.CreationTimestamp,
+		}
+		return nil
+	}
+
+	if policies, exist := mgr.resPolicies[resKey]; exist {
+		// check if it needs update
+		if !policiesConfigmap.CreationTimestamp.Equal(&policies.configCreationTimestamp) {
+			if err := updateResPolicies(); err != nil {
+				return err
+			}
+		}
+	} else {
+		if err := updateResPolicies(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (mgr *ResPoliciesManager) GetStructredVolumeForPVC(namespace, name string) (*StructuredVolume, error) {
+	pvc := &v1.PersistentVolumeClaim{}
+	err := mgr.kbClient.Get(context.Background(), kbClient.ObjectKey{Namespace: namespace, Name: name}, pvc)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	if pvc.Spec.VolumeName == "" {
+		return nil, errors.Errorf("PVC %s/%s has no volume backing this claim", namespace, name)
+	}
+	if pvc.Status.Phase != v1.ClaimBound {
+		return nil, errors.Errorf("PVC %s/%s is in phase %v and is not bound to a volume", namespace, name, pvc.Status.Phase)
+	}
+	pvName := pvc.Spec.VolumeName
+	pv := &v1.PersistentVolume{}
+	err = mgr.kbClient.Get(context.Background(), kbClient.ObjectKey{Name: pvName}, pv)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	volume := StructuredVolume{}
+	volume.ParsePV(pv)
+	return &volume, nil
+}
+
+func (mgr *ResPoliciesManager) GetVolumeMatchedAction(refName, refType string, volume *StructuredVolume) (*Action, error) {
```

Comment:
yes, this GetVolumeMatchedAction is called in backup process.
the handle logic entry similar like below

```golang
func (ib *itemBackupper) executeActions(...)
.....
	if groupResource.String() == "persistentvolumeclaims" && ib.backupRequest.Spec.ResourcePolices != nil {
	    volume, err := respolicies.ResPoliciesMgr.GetStructredVolumeForPVC(namespace, name)
	   action, err := respolicies.ResPoliciesMgr.GetVolumeMatchedAction(ib.backupRequest.Spec.ResourcePolices.RefName, ib.backupRequest.Spec.ResourcePolices.RefType, volume)
         if action != nil && action.Type == respolicies.Skip {
....
}
```

Reply:
The `configmap` should be read once for one backup and the `ResourcePolicies` should be built once as well for one backup and immutable during the whole backup processing.

Reply:
yes, currently, the `ResourcePolicies` only build once if it's up-to-date.

the `ResourcePolicies` is stored in `resPoliciesMeta` and can only be updated when doing backup processing. and backup processing is related to one specific backup CR in reconciliation, So it could not be updated in parallel.